### PR TITLE
github_runner_matrix: deploy new x86_64 runner when required

### LINKED ISSUE #18356: Setup `brew determine-test-runners` for macOS 15 x86_64
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

We're ready to enable macOS 15 x86_64 for PRs but the setup is a little more complex.

We need `brew determine-test-runners` to add macOS 15 x86_64 to the matrix if any of the following are true:

* The package is `pkg-config` or `pkgconf`
* `formula.class.pour_bottle_only_if == :clt_installed`
* Any dependent of the tested formula satisfies `dependent_formula.class.pour_bottle_only_if == :clt_installed`

The last condition may require making `--eval-all` mandatory.

### What is the motivation for the feature?

In order to have proper testing of macOS 15 x86_64 and not require manual post-merge building.

### How will the feature be relevant to at least 90% of Homebrew users?

They will get upgraded to compatible bottles rather than temporarily get incompatible bottles that break their workflow.

### What alternatives to the feature have been considered?

None

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Closes #18356.

### CODE REVIEW

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks @carlocab, great work so far!

File: Library/Homebrew/github_runner_matrix.rb [hunk 1]
```diff
@@ -129,6 +129,23 @@ def runner_enabled?(macos_version)
...
+  NEW_INTEL_MACOS_MUST_BUILD_FORMULAE = %w[pkg-config pkgconf].freeze
+
+  sig { returns(T::Boolean) }
+  def deploy_new_x86_64_runner?
+    return true if @testing_formulae.any? { |f| NEW_INTEL_MACOS_MUST_BUILD_FORMULAE.include?(f.name) }
+    return true if @testing_formulae.any? { |f| f.formula.class.pour_bottle_only_if == :clt_installed }
+
+    Formula.all.any? do |formula|
+      non_test_dependencies = Dependency.expand(formula, cache_key: "determine-test-runners") do |_, dependency|
+        Dependency.prune if dependency.test?
```
Comment (Bo98) [inline Library/Homebrew/github_runner_matrix.rb:141]: I don't think the unbottled check works this way. Even test dependencies get bottled (though there's basically no test-only dependencies in the affected trees anyway).

File: Library/Homebrew/github_runner_matrix.rb [hunk 2]
```diff
@@ -129,6 +129,23 @@ def runner_enabled?(macos_version)
...
+      end
+      next false if non_test_dependencies.none? { |dep| @testing_formulae.map(&:name).include?(dep.name) }
+
+      formula.class.pour_bottle_only_if == :clt_installed
```
Comment (Bo98) [inline Library/Homebrew/github_runner_matrix.rb:145]: I'd move this before the dependency check (`return false if formula.class.pour_bottle_only_if != :clt_installed`). This should make it significantly faster.

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/github_runner_matrix.rb (hunk 1 above)
Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:141]: It's how it's meant to work. If that's not how it actually works then that's a different story:

https://github.com/Homebrew/homebrew-test-bot/blob/e84b3ddb0821cd0145f1b8648705813df3c8e2bd/lib/tests/formulae.rb#L304-L316

File: Library/Homebrew/github_runner_matrix.rb (hunk 2 above)
Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:145]: Are you referring to

https://github.com/Homebrew/brew/blob/41c5f7243f6813b17d8f05c2f862c5869049d04d/Library/Homebrew/github_runner_matrix.rb#L137

? Or am I misunderstanding you?

Comment (Bo98) [reply Library/Homebrew/github_runner_matrix.rb:145]: No, within the `Formula.all?` block.

File: Library/Homebrew/github_runner_matrix.rb (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/github_runner_matrix.rb:141]: Ah I see what mistake I made - didn't realise `libgccjit` has a Intel only dependency.

Which is weird but whatever.

File: Library/Homebrew/github_runner_matrix.rb [hunk 3]
```diff
@@ -129,6 +129,23 @@ def runner_enabled?(macos_version)
...
+    end
+  end
```
Comment (Bo98) [inline Library/Homebrew/github_runner_matrix.rb:147]:
```suggestion
    Formula.all.any? do |formula|
      return false if formula.class.pour_bottle_only_if != :clt_installed
    
      non_test_dependencies = Dependency.expand(formula, cache_key: "determine-test-runners") do |_, dependency|
        Dependency.prune if dependency.test?
      end
      non_test_dependencies.any? { |dep| @testing_formulae.map(&:name).include?(dep.name) }
    end
  end
```

File: Library/Homebrew/github_runner_matrix.rb (hunk 2 above)
Comment (Bo98) [reply Library/Homebrew/github_runner_matrix.rb:145]: Added a suggestion diff though might be hard to read.

File: Library/Homebrew/github_runner_matrix.rb (hunk 3 above)
Comment (Bo98) [reply Library/Homebrew/github_runner_matrix.rb:147]: yes, fixed

Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:147]: Ok, the `return false` is what threw me off when you initially suggested it. Thanks!

File: Library/Homebrew/github_runner_matrix.rb [hunk 4]
```diff
@@ -129,6 +129,23 @@ def runner_enabled?(macos_version)
...
   end
 
+  NEW_INTEL_MACOS_MUST_BUILD_FORMULAE = %w[pkg-config pkgconf].freeze
+
+  sig { returns(T::Boolean) }
+  def deploy_new_x86_64_runner?
+    return true if @testing_formulae.any? { |f| NEW_INTEL_MACOS_MUST_BUILD_FORMULAE.include?(f.name) }
+    return true if @testing_formulae.any? { |f| f.formula.class.pour_bottle_only_if == :clt_installed }
+
+    Formula.all.any? do |formula|
```
[the remark this answers is not in the document]
Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:140]: I can skip formulae not in the given tap, but `Tap` objects don't seem to have a nice way to recover the formulae they contain.

File: Library/Homebrew/github_runner_matrix.rb [hunk 5]
```diff
@@ -122,13 +123,30 @@ def create_runner(platform, arch, spec, macos_version = nil)
...
```
Comment (Bo98) [inline Library/Homebrew/github_runner_matrix.rb:140]: You probably need to pass `eval_all` here

File: Library/Homebrew/github_runner_matrix.rb [hunk 6]
```diff
@@ -181,7 +199,13 @@ def generate_runners!
       )
       @runners << create_runner(:macos, :arm64, spec, macos_version)
 
-      next if !@all_supported && macos_version > NEWEST_HOMEBREW_CORE_INTEL_MACOS_RUNNER
+      # `#deploy_new_x86_64_runner?` is expensive, so:
+      #  - avoid calling it if we don't have to
+      #  - cache the result to a variable to avoid calling it multiple times
+      if macos_version > NEWEST_DEFAULT_HOMEBREW_CORE_INTEL_MACOS_RUNNER &&
```
Comment (Bo98) [inline Library/Homebrew/github_runner_matrix.rb:205]: Was removing the `!@all_supported` intentional?

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/github_runner_matrix.rb (hunk 6 above)
Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:205]: It's up here: https://github.com/Homebrew/brew/blob/30e25c540372fd0394ee2c8e0dbf06651cc125f1/Library/Homebrew/github_runner_matrix.rb#L63

File: Library/Homebrew/github_runner_matrix.rb (hunk 5 above)
Comment (carlocab) [reply Library/Homebrew/github_runner_matrix.rb:140]: Yep, will add it when I get around to updating the calls in `homebrew/core`

File: Library/Homebrew/github_runner_matrix.rb (hunk 6 above)
Comment (Bo98) [reply Library/Homebrew/github_runner_matrix.rb:205]: Ah thanks, that wasn't obvious.

File: Library/Homebrew/github_runner_matrix.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/github_runner_matrix.rb:140]: I think handling formulae files is always nicer than doing `Formula.all` and evaluating every formula, whether needed or not.

**Comment by carlocab** [comment]
I'll get to this late next week (or in the first half of the week following that).

**Pull request closed without merging**